File: electron/database/noteConnectionRepository.ts
import type Database from "better-sqlite3";
import type { NoteConnection } from "../types";

interface NoteConnectionRow {
  id: string;
  source_note_id: string;
  target_note_id: string;
  created_at: string;
}

function rowToConnection(row: NoteConnectionRow): NoteConnection {
  return {
    id: row.id,
    sourceNoteId: row.source_note_id,
    targetNoteId: row.target_note_id,
    createdAt: row.created_at,
  };
}

export function createNoteConnectionRepository(db: Database.Database) {
  const stmts = {
    fetchAll: db.prepare(
      `SELECT * FROM note_connections ORDER BY created_at ASC`,
    ),
    fetchById: db.prepare(`SELECT * FROM note_connections WHERE id = ?`),
    insert: db.prepare(`
      INSERT INTO note_connections (id, source_note_id, target_note_id, created_at)
      VALUES (@id, @source_note_id, @target_note_id, @created_at)
    `),
    delete: db.prepare(`DELETE FROM note_connections WHERE id = ?`),
    deleteByNotePair: db.prepare(`
      DELETE FROM note_connections
      WHERE (source_note_id = @source AND target_note_id = @target)
         OR (source_note_id = @target AND target_note_id = @source)
    `),
  };

  return {
    fetchAll(): NoteConnection[] {
      return (stmts.fetchAll.all() as NoteConnectionRow[]).map(rowToConnection);
    },

    create(sourceNoteId: string, targetNoteId: string): NoteConnection {
      const now = new Date().toISOString();
      const id = `nc-${crypto.randomUUID()}`;
      stmts.insert.run({
        id,
        source_note_id: sourceNoteId,
        target_note_id: targetNoteId,
        created_at: now,
      });
      const row = stmts.fetchById.get(id) as NoteConnectionRow;
      return rowToConnection(row);
    },

    delete(id: string): void {
      stmts.delete.run(id);
    },

    deleteByNotePair(sourceNoteId: string, targetNoteId: string): void {
      stmts.deleteByNotePair.run({
        source: sourceNoteId,
        target: targetNoteId,
      });
    },
  };
}

export type NoteConnectionRepository = ReturnType<
  typeof createNoteConnectionRepository
>;

File: src-tauri/src/db/note_connection_repository.rs
use rusqlite::{params, Connection};
use serde::{Deserialize, Serialize};

use super::helpers;
use super::row_converter::{query_all, query_one, FromRow};

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct NoteConnection {
    pub id: String,
    pub source_note_id: String,
    pub target_note_id: String,
    pub created_at: String,
    pub updated_at: String,
}

impl FromRow for NoteConnection {
    fn from_row(row: &rusqlite::Row) -> rusqlite::Result<Self> {
        Ok(NoteConnection {
            id: row.get("id")?,
            source_note_id: row.get("source_note_id")?,
            target_note_id: row.get("target_note_id")?,
            created_at: row.get("created_at")?,
            updated_at: row.get("updated_at")?,
        })
    }
}

pub fn fetch_all(conn: &Connection) -> rusqlite::Result<Vec<NoteConnection>> {
    query_all(
        conn,
        "SELECT * FROM note_connections ORDER BY created_at",
        [],
    )
}

pub fn create(
    conn: &Connection,
    source_note_id: &str,
    target_note_id: &str,
) -> rusqlite::Result<NoteConnection> {
    let id = format!("nc-{}", helpers::new_uuid());
    let now = helpers::now();
    conn.execute(
        "INSERT INTO note_connections (id, source_note_id, target_note_id, created_at, updated_at) \
         VALUES (?1, ?2, ?3, ?4, ?5)",
        params![&id, source_note_id, target_note_id, &now, &now],
    )?;

    query_one(conn, "SELECT * FROM note_connections WHERE id = ?1", [&id])
}

pub fn delete(conn: &Connection, id: &str) -> rusqlite::Result<()> {
    conn.execute("DELETE FROM note_connections WHERE id = ?1", [id])?;
    Ok(())
}

pub fn delete_by_note_pair(
    conn: &Connection,
    source_note_id: &str,
    target_note_id: &str,
) -> rusqlite::Result<()> {
    conn.execute(
        "DELETE FROM note_connections \
         WHERE (source_note_id = ?1 AND target_note_id = ?2) \
         OR (source_note_id = ?2 AND target_note_id = ?1)",
        params![source_note_id, target_note_id],
    )?;
    Ok(())
}
